design/secrets_provider_milestone1_design.md

```diff
@@ -0,0 +1,435 @@
+During Phase 1, an *init container* was deployed in the same Pod as the customer's applications and populate the K8s secrets defined in the customer's manifest.
+
+### Motivation
+
+The current implementation of the Secrets Provider only supports one app and makes the deployment of the Secrets Provider deeply coupled to the customer's applications.
+Also, it launches once and does not support secret value rotation while the app is running.
+
+### Requirements
+
+The solution should support the following capabilities:
+
+Milestone 1
+
+- Secrets Provider runs and deployed without affecting the app deployment or app life cycle
+- Secrets Provider can serve many apps
+
+Milestone 2
+
+- Secrets Provider support rotations
+
+### Future requirements
+
+The solution should consider the following future requirements that may arise:
+
+* Support other **targets** for the secrets besides K8s Secrets, such as files
+* Support other **sources** for secrets paths besides conjur-map in K8s Secrets
+* Support **sources** varying while Secrets Provider runs
+
+## Solution
+The solution is to allow different deployments and behave differently based on the deployment chosen.
+
+The following image shows the desicion flow chart that the customer will use to determine how to configure Secrets Provider depends on the value he wants to get:
+
+![Secrets Provider flavors desicion flow chart](https://user-images.githubusercontent.com/31179631/85747023-975bf500-b70f-11ea-8e26-1134068fe655.png)
+
+This flow chart shows 6 different variations of Secrets Provider:
+
+| Secrets Provider variation                            | Business value                                               | Implementation status |
+| ----------------------------------------------------- | ------------------------------------------------------------ | --------------------- |
+| Run as **Init Container** and sync to **K8s Secrets** | Supply secrets on application load, based on K8s secrets     | Implemented           |
+| Run as **Job** and sync to **K8s Secrets**            | Supply secrets once for many applications, based on K8s secrets | Milestone 1           |
+| Run as **Deployment** and sync to **K8s Secrets**     | Supply secrets for many applications and support rotation, based on K8s secrets | Future (Milestone 2)  |
+| Run as **Sidecar** and sync to **K8s Secrets**        | Support secrets rotation, based on K8s secrets               | Future (Milestone 2)  |
+| Run as **Init Container** and sync to **files**       | Supply secrets on application load, pushed as files to shared volume | Future (Milestone 3)  |
+| Run as **Sidecar** and sync to **files**              | Supply secrets for one application and support rotation, pushed as files to shared volume | Future (Milestone 3)  |
+
+All these variations will be supported using the same Secrets Provider image, and the behavior will vary dynamically depends on the chosen deployment.
```

Comment:
What's the reason for that? Is it really better to encapsulate all this functionality into a single image?

Reply:
It is easier to maintain a single product and not multiple forks with slight variation between them.
We plan to do it modularly, so it can be pulled out if necessary.